pyop2/fusion.py

```diff
@@ -0,0 +1,1264 @@
+
+    @classmethod
+    def _cache_key(cls, kernels, fused_ast=None, loop_chain_index=None):
+        keys = "".join([super(Kernel, cls)._cache_key(k._code or k._ast.gencode(),
+                                                      k._name, k._opts, k._include_dirs,
+                                                      k._headers, k._user_code)
+                        for k in kernels])
+        return str(loop_chain_index) + keys
+
+    def _ast_to_c(self, asts, opts):
+        """Fuse Abstract Syntax Trees of a collection of kernels and transform
+        them into a string of C code."""
+        if not isinstance(asts, (ast.FunDecl, ast.Root)):
+            asts = ast.Root(asts)
+        self._ast = asts
+        self._original_ast = dcopy(asts)
+        return super(Kernel, self)._ast_to_c(self._ast, opts)
+
+    def __init__(self, kernels, fused_ast=None, loop_chain_index=None):
+        """Initialize a :class:`fusion.Kernel` object.
+
+        :param kernels: an iterator of some :class:`Kernel` objects. The objects
+                        can be of class `fusion.Kernel` or of any superclass.
+        :param fused_ast: the Abstract Syntax Tree of the fused kernel. If not
+                          provided, kernels are simply concatenated.
+        :param loop_chain_index: index (i.e., position) of the kernel in a loop
+                                 chain. This can be used to differentiate a same
+                                 kernel appearing multiple times in a loop chain.
+        """
+        # Protect against re-initialization when retrieved from cache
+        if self._initialized:
+            return
+
+        asts = fused_ast
+        if not asts:
+            # If kernels' need be concatenated, discard duplicates
+            kernels = dict(zip([k.cache_key[1:] for k in kernels], kernels)).values()
+            asts = [k._ast for k in kernels]
+        kernels = as_tuple(kernels, (Kernel, host.Kernel, base.Kernel))
+
+        Kernel._globalcount += 1
+        self._kernels = kernels
+        self._name = "_".join([k.name for k in kernels])
+        self._opts = dict(flatten([k._opts.items() for k in kernels]))
+        self._applied_blas = any(k._applied_blas for k in kernels)
+        self._include_dirs = list(set(flatten([k._include_dirs for k in kernels])))
+        self._headers = list(set(flatten([k._headers for k in kernels])))
+        self._user_code = "\n".join(list(set([k._user_code for k in kernels])))
+
+        # Code generation is delayed until actually needed
+        self._ast = asts
+        self._code = None
+
+        self._initialized = True
+
+    def __iter__(self):
+        for k in self._kernels:
+            yield k
+
+    def __str__(self):
+        return "OP2 FusionKernel: %s" % self._name
+
+
+# Parallel loop API
+
+class JITModule(host.JITModule):
+
+    _cppargs = []
+    _libraries = []
+    _extension = 'cpp'
+
+    _wrapper = """
+extern "C" void %(wrapper_name)s(%(executor_arg)s,
```

Comment:
Is this template really new? Don't we have it somewhere in the software stack already? Just asking to avoid potential duplication...